testproxy/services/check-and-mutate-row.js

```diff
@@ -16,30 +16,89 @@
 
 const normalizeCallback = require('./utils/normalize-callback.js');
+const getTableInfo = require('./utils/get-table-info');
+const {
+  createFlatMutationsListWithFnInverse,
+} = require('../../build/testproxy/services/utils/request/createFlatMutationsList');
+const {
+  mutationParseInverse,
+} = require('../../build/testproxy/services/utils/request/mutateInverse');
 
-const v2 = Symbol.for('v2');
+/**
+ * Transforms mutations from the gRPC layer format to the handwritten layer format.
+ * This function takes an array of gRPC layer mutations (google.bigtable.v2.IMutation[]) and converts
+ * them back to the format used by the handwritten layer.  It essentially reverses the transformation
+ * performed by Mutation.parse.  It's used internally by the test proxy for checkAndMutateRow.
+ *
+ * @param {google.bigtable.v2.IMutation[]} gapicLayerMutations An array of mutations in the gRPC layer format.
+ * @returns {FilterConfigOption[]} An array of mutations in the handwritten layer format.
+ */
+function handwrittenLayerMutations(gapicLayerMutations) {
+  return createFlatMutationsListWithFnInverse(
+    [
+      {
+        mutations: gapicLayerMutations,
+      },
+    ],
+    mutationParseInverse,
+    1
+  );
+}
+
+/**
+ * Converts a byte array (or string) back to a string.  This is the inverse of
+ * Mutation.convertToBytes.
+ *
+ * @param {Bytes} bytes The byte array or string to convert.
+ * @returns {string} The converted string.
+ */
+function convertFromBytes(bytes) {
+  if (bytes instanceof Buffer) {
+    return bytes.toString();
+  } else if (typeof bytes === 'string') {
+    return bytes;
+  } else {
+    throw new Error('Invalid input type. Must be Buffer or string.');
+  }
+}
 
 const checkAndMutateRow = ({clientMap}) =>
   normalizeCallback(async rawRequest => {
     const {request} = rawRequest;
-    const {request: checkAndMutateRequest} = request;
-    const {appProfileId, falseMutations, rowKey, tableName, trueMutations} =
-      checkAndMutateRequest;
-
-    const {clientId} = request;
-    const client = clientMap.get(clientId)[v2];
-    const [result] = await client.checkAndMutateRow({
+    const {clientId, request: checkAndMutateRowRequest} = request;
+    const {
+      // authorizedViewName, // TODO: Pass the authorizedViewName along in the test proxy.
       appProfileId,
       falseMutations,
       rowKey,
+      // predicateFilter, // TODO: Pass the predicateFilter along in the test proxy.
       tableName,
       trueMutations,
-    });
-
-    return {
-      status: {code: grpc.status.OK, details: []},
-      result,
-    };
+    } = checkAndMutateRowRequest;
+    const onMatch = handwrittenLayerMutations(trueMutations);
+    const onNoMatch = handwrittenLayerMutations(falseMutations);
+    const id = convertFromBytes(rowKey);
+    const bigtable = clientMap.get(clientId);
+    bigtable.appProfileId = appProfileId; // TODO: Remove this line and pass appProfileId into row.filter.
```

Comment:
is todo needed?

Reply:
Yes. The `filter` method currently does not accept an `appProfileId`, but we should allow the app profile id to be overridden per method call. The evidence of this is that the `appProfileId` is passed into the `checkAndMutateService` and not the `createClient` service so the intent is to set the appProfileId per method call.

Reply:
Remove TODO